service/tcp.go

```diff
@@ -207,121 +208,125 @@ func (s *tcpService) Serve(listener *net.TCPListener) error {
-	clientTCPConn.SetReadDeadline(connStart.Add(s.readTimeout))
-	var proxyMetrics metrics.ProxyMetrics
-	clientConn := metrics.MeasureConn(clientTCPConn, &proxyMetrics.ProxyClient, &proxyMetrics.ClientProxy)
-	cipherEntry, clientReader, clientSalt, timeToCipher, keyErr := findAccessKey(clientConn, remoteIP(clientTCPConn), s.ciphers)
-
-	connError := func() *onet.ConnectionError {
-		if keyErr != nil {
-			logger.Debugf("Failed to find a valid cipher after reading %v bytes: %v", proxyMetrics.ClientProxy, keyErr)
-			const status = "ERR_CIPHER"
-			s.absorbProbe(listenerPort, clientConn, clientLocation, status, &proxyMetrics)
-			return onet.NewConnectionError(status, "Failed to find a valid cipher", keyErr)
-		}
-
-		isServerSalt := cipherEntry.SaltGenerator.IsServerSalt(clientSalt)
-		// Only check the cache if findAccessKey succeeded and the salt is unrecognized.
-		if isServerSalt || !s.replayCache.Add(cipherEntry.ID, clientSalt) {
-			var status string
-			if isServerSalt {
-				status = "ERR_REPLAY_SERVER"
-			} else {
-				status = "ERR_REPLAY_CLIENT"
-			}
-			s.absorbProbe(listenerPort, clientConn, clientLocation, status, &proxyMetrics)
-			logger.Debugf(status+": %v in %s sent %d bytes", clientTCPConn.RemoteAddr(), clientLocation, proxyMetrics.ClientProxy)
-			return onet.NewConnectionError(status, "Replay detected", nil)
-		}
-
-		ssr := ss.NewShadowsocksReader(clientReader, cipherEntry.Cipher)
-		tgtAddr, err := socks.ReadAddr(ssr)
-		// Clear the deadline for the target address
-		clientTCPConn.SetReadDeadline(time.Time{})
-		if err != nil {
-			// Drain to prevent a close on cipher error.
-			io.Copy(ioutil.Discard, clientConn)
-			return onet.NewConnectionError("ERR_READ_ADDRESS", "Failed to get target address", err)
-		}
+	clientConn.SetReadDeadline(time.Now().Add(s.readTimeout))
+
+	// 1. Find the cipher and acess key id.
+	cipherEntry, clientReader, clientSalt, timeToCipher, keyErr := findAccessKey(clientConn, remoteIP(clientConn), s.ciphers)
+	s.m.AddTCPCipherSearch(keyErr == nil, timeToCipher)
+	if keyErr != nil {
+		logger.Debugf("Failed to find a valid cipher after reading %v bytes: %v", proxyMetrics.ClientProxy, keyErr)
+		const status = "ERR_CIPHER"
+		s.absorbProbe(listenerPort, clientConn, status, proxyMetrics)
+		return "", onet.NewConnectionError(status, "Failed to find a valid cipher", keyErr)
+	}
+	var id string
+	if cipherEntry != nil {
+		id = cipherEntry.ID
+	}
 
-		tgtConn, dialErr := dialTarget(tgtAddr, &proxyMetrics, s.targetIPValidator)
-		if dialErr != nil {
-			// We don't drain so dial errors and invalid addresses are communicated quickly.
-			return dialErr
+	// 2. Check if the connection is a replay.
+	isServerSalt := cipherEntry.SaltGenerator.IsServerSalt(clientSalt)
+	// Only check the cache if findAccessKey succeeded and the salt is unrecognized.
+	if isServerSalt || !s.replayCache.Add(cipherEntry.ID, clientSalt) {
+		var status string
+		if isServerSalt {
+			status = "ERR_REPLAY_SERVER"
```

Comment:
Are these error status used by client? If so, shall we use a dedicated file listing all these constants?

Reply:
This is used in the prometheus metrics, as a dimension. I think we could define constants, but perhaps that can be part of an error revamp, so I'd rather punt that to later.